{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d5c932-7752-4f1f-87ce-002e560b56b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "\n",
    "# See \"Diseases.ipynb\" for these files\n",
    "custom = pd.read_excel('./Phenotyping/custom_phenotypes.xlsx')\n",
    "custom['phecode1.2'] = custom['phecode1.2'].astype(float)\n",
    "icd = pd.read_csv('./Phenotyping/icd_codes.csv')\n",
    "merged_pheno = pd.read_excel('./Phenotyping/merged_phenotypes.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "937653d4-e85d-4583-a02f-0da67351ee46",
   "metadata": {},
   "source": [
    "## MVP"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1092ccde-a115-4a2f-b7e8-7fe2ffaa2a47",
   "metadata": {},
   "source": [
    "# Download files in parallel\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_CirculatorySystem_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_CirculatorySystem_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_CirculatorySystem_batch3.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_CongenitalAnomalies.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_DermatologicalSystem.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_DigestiveSystem_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_DigestiveSystem_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_EndocrineMetabolic_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_EndocrineMetabolic_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Genitourinary_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Genitourinary_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Hematopoietic.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_MusculoskeletalSystem_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_MusculoskeletalSystem_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Neurological.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Other.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Pregnancy.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_RespiratorySystem.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_SenseOrgans_batch1.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_SenseOrgans_batch2.analysis-PI.MULTI.tar\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/MVP https://ftp.ncbi.nlm.nih.gov/dbgap/studies/phs002453/analyses/GIA/phs002453.MVP_R4.1000G_AGR.GIA.PheCodes_Symptoms.analysis-PI.MULTI.tar\"\n",
    "\n",
    "# Extract files in parallel\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o extract.out -e extract.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" bash -c 'mkdir -p /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted && for tar_file in /sc/arion/projects/GENECAD/Robert/DOE/MVP/*CirculatorySystem*.tar; do tar --strip-components=1 -xf \"$tar_file\" -C /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted; done'\n",
    "\n",
    "... same for other phenotypes\n",
    "\n",
    "# Extract and clean files (keep only variants with p < 0.05)\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o cleanup.out -e cleanup.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" 'find /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted -type f ! \\( -name \"*META*\" -a -name \"*.txt.gz\" \\) -delete'\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o cleanup.out -e cleanup.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" \\\n",
    "'for f in /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted/*.txt.gz; do \\\n",
    "mv \"$f\" \"/sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted/$(basename \"$f\" | sed \"s/^MVP_R4\\.1000G_AGR\\.//;s/\\.META\\.GIA\\.dbGaP//\")\"; \\\n",
    "done'\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 8 -o clean.out -e clean.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" 'for f in /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted/*.txt.gz; do b=$(basename \"$f\" .txt.gz); pigz -dc \"$f\" | awk -v p=\"$b\" '\\''NR==1{print \"chrom\",\"pos\",\"ref\",\"alt\",\"or\",\"pval\",\"pheno\"} NR>1 && $15<0.05{print $2,$3,$4,$5,$13,$15,p}'\\'' > /sc/arion/projects/GENECAD/Robert/DOE/MVP/Cleaned/\"$b\".txt; done'\n",
    "\n",
    "# Get unique variants from sampling of phenotypes\n",
    "\n",
    "echo -e \"chrom\\tpos\\tref\\talt\" > /sc/arion/projects/GENECAD/Robert/DOE/MVP/unique_variants.txt\n",
    "for f in Phe_374_6.txt.gz Phe_521_2.txt.gz Phe_706_1.txt.gz Phe_696_41.txt.gz Phe_696_4.txt.gz Phe_370.txt.gz Phe_696.txt.gz Phe_274_11.txt.gz Phe_371_21.txt.gz Phe_372.txt.gz\n",
    "do\n",
    "pigz -dc /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted/\"$f\" | awk 'NR>1{print $2,$3,$4,$5}'\n",
    "done | sort -u >> /sc/arion/projects/GENECAD/Robert/DOE/MVP/unique_variants.txt\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 8 -o unique.out -e unique.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" 'echo -e \"chrom\\tpos\\tref\\talt\" > /sc/arion/projects/GENECAD/Robert/DOE/MVP/unique_variants.txt; for f in Phe_374_6.txt.gz Phe_521_2.txt.gz Phe_706_1.txt.gz Phe_696_41.txt.gz Phe_696_4.txt.gz Phe_370.txt.gz Phe_696.txt.gz Phe_274_11.txt.gz Phe_371_21.txt.gz Phe_372.txt.gz; do pigz -dc /sc/arion/projects/GENECAD/Robert/DOE/MVP/Extracted/\"$f\" | awk '\\''NR>1{print $2,$3,$4,$5}'\\''; done | sort -u >> /sc/arion/projects/GENECAD/Robert/DOE/MVP/unique_variants.txt'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef7b7658-38d8-497a-89cb-a82ccd15cc51",
   "metadata": {},
   "source": [
    "## Genebass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2711982-1788-4a4a-8569-aecb417f3b01",
   "metadata": {},
   "source": [
    "### Exporting data with Hail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d568a0fb-5d78-483c-9369-a1acb0367679",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import hail as hl\n",
    "hl.init(min_block_size=128)\n",
    "import pandas as pd\n",
    "!export PYSPARK_SUBMIT_ARGS=\"--driver-memory 144g --executor-memory 144g pyspark-shell\"\n",
    "\n",
    "pt = hl.read_table('/sc/arion/projects/data-ark/Public_Unrestricted/Genebass/pheno_results.ht')\n",
    "df = pt.to_pandas()\n",
    "df.to_excel('./Genebass/pheno_results.xlsx')\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "13affc0e-2370-4c4a-9ba0-5cceb52a64e1",
   "metadata": {},
   "source": [
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies && wget -P /sc/arion/projects/GENECAD/Robert/dbNSFP https://dbnsfp.s3.amazonaws.com/dbNSFP4.9a.zip\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 6 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=2GB] \"unzip /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP4.9a.zip\"\n",
    "\n",
    "version=4.9a\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 12 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=2GB] \"ml tabix && zgrep -h -v ^#chr /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_variant.chr* | sort -k1,1 -k2,2n - | cat h - | bgzip -c > /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_grch38.gz && tabix -s 1 -b 2 -e 2 /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_grch38.gz\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 12 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=2GB] \"ml tabix && zgrep -h -v ^#chr /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_variant.chr* | awk '\\$8 != \\\".\\\" ' | sort -k8,8 -k9,9n - | cat h - | bgzip -c > /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_grch37.gz && tabix -s 8 -b 9 -e 9 /sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP${version}_grch37.gz\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5fee18fd-4505-4ced-a3e4-0a044ee00c5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLF4J: Failed to load class \"org.slf4j.impl.StaticMDCBinder\".\n",
      "SLF4J: Defaulting to no-operation MDCAdapter implementation.\n",
      "SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.\n",
      "2024-12-14 14:44:20.249 Hail: INFO: merging 1001 files totalling 374.9M... 1000]\n",
      "2024-12-14 14:44:20.672 Hail: INFO: while writing:\n",
      "    /sc/arion/projects/GENECAD/Robert/DOE/Genebass/GB/gb.tsv\n",
      "  merge time: 422.232ms\n"
     ]
    }
   ],
   "source": [
    "gb = hl.read_matrix_table('/sc/arion/projects/data-ark/Public_Unrestricted/Genebass/results.mt')\n",
    "gb = gb.drop('markerAFs','n_cases','n_controls','heritability','saige_version','inv_normalized','n_cases_defined',\n",
    "             'n_cases_both_sexes','n_cases_males','n_cases_females','coding_description','description_more',\n",
    "             'interval','markerIDs','Nmarker_MACCate_1','Nmarker_MACCate_2','Nmarker_MACCate_3',\n",
    "             'Nmarker_MACCate_4','Nmarker_MACCate_5','Nmarker_MACCate_6','Nmarker_MACCate_7','Nmarker_MACCate_8',\n",
    "             'Pvalue.NA','Pvalue_Burden.NA','Pvalue_SKAT.NA','BETA_Burden.NA','SE_Burden.NA',\n",
    "             'Pvalue','Pvalue_SKAT')\n",
    "gb = gb.filter_cols(\n",
    "    (gb.trait_type == 'icd10') | (gb.trait_type == 'icd_first_occurrence') | ((gb.modifier == 'custom') & (gb.trait_type == 'categorical'))\n",
    ")\n",
    "gb = gb.filter_rows((gb.annotation != 'synonymous'))\n",
    "\n",
    "gb = gb.filter_entries((gb.Pvalue_Burden < 0.05))\n",
    "\n",
    "gb = gb.key_rows_by()\n",
    "gb = gb.key_cols_by() \n",
    "gb = gb.drop('pheno_sex','trait_type','coding','modifier')\n",
    "\n",
    "gb = gb.entries()\n",
    "gb.export('/sc/arion/projects/GENECAD/Robert/DOE/Genebass/GB/gb.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67452c00-59b7-4c07-9adf-ff9952610e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = hl.read_matrix_table('/sc/arion/projects/data-ark/Public_Unrestricted/Genebass/variant_results.mt')\n",
    "mt = mt.drop('heritability','saige_version','inv_normalized',\n",
    "             'n_cases_both_sexes','n_cases_females','n_cases_males','coding_description','description_more',\n",
    "             'coding_description','description','n_cases_defined','category',\n",
    "             'call_stats','AF.Cases','AF.Controls')\n",
    "mt = mt.filter_cols(\n",
    "    (mt.trait_type == 'icd10') | (mt.trait_type == 'icd_first_occurrence') | ((mt.modifier == 'custom') & (mt.trait_type == 'categorical'))\n",
    ")\n",
    "mt = mt.filter_rows((mt.annotation != 'synonymous'))\n",
    "\n",
    "mt = mt.filter_entries((mt.Pvalue < 0.05))\n",
    "\n",
    "mt = mt.key_rows_by()\n",
    "mt = mt.key_cols_by() \n",
    "mt = mt.drop('locus','alleles','pheno_sex','trait_type','coding','modifier')\n",
    "\n",
    "mt = mt.entries()\n",
    "mt.export('/sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv.tsv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed7bd681-c32f-41cb-b527-b99e0b416234",
   "metadata": {},
   "source": [
    "### Annotating variants with VEP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a78277-2d6b-4334-8988-0d117e85fa9b",
   "metadata": {},
   "source": [
    "#### Running VEP with missense scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e297ae8-b1e9-4037-8679-a7f2a233f740",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv.tsv', sep='\\t')\n",
    "df.to_pickle('./Genebass/SV/sv.pkl')\n",
    "df['markerID'].drop_duplicates().to_pickle('./Genebass/SV/sv_unique.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ad7c04-5cc8-4feb-8206-18ab0867c0f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle('./Genebass/SV/sv_unique.pkl')\n",
    "df = pd.DataFrame(df)\n",
    "df['chr'] = df['markerID'].str.split(':').str[0]\n",
    "df['pos'] = df['markerID'].str.split(':').str[1].str.split('_')\n",
    "df['ref'] = df['pos'].str[1]\n",
    "df['pos'] = df['pos'].str[0].astype(int)\n",
    "df['ref'] = df['ref'].str.split('/')\n",
    "df['alt'] = df['ref'].str[1]\n",
    "df['ref'] = df['ref'].str[0]\n",
    "df['chr'] = df['chr'].str.replace('chr','')\n",
    "df = df.reset_index(drop=True)\n",
    "df = df.rename({'markerID':'id'},axis=1)\n",
    "\n",
    "# Add necessary VCF fields\n",
    "df['qual'] = '.'\n",
    "df['filter'] = '.'\n",
    "df['info'] = '.'\n",
    "\n",
    "# Reorder columns to VCF format\n",
    "vcf_columns = ['chr', 'pos', 'id', 'ref', 'alt', 'qual', 'filter', 'info']\n",
    "vcf_df = df[vcf_columns]\n",
    "vcf_df = vcf_df.sort_values(['chr','pos','ref','alt'])\n",
    "\n",
    "# Save as VCF in chunks for VEP annotation\n",
    "vcf_header = \"\"\"##fileformat=VCFv4.2\n",
    "##source=CustomGeneratedVCF\n",
    "#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\n",
    "\"\"\"\n",
    "with open('./Genebass/SV/sv_unique.vcf', 'w') as f:\n",
    "    f.write(vcf_header)\n",
    "    vcf_df.to_csv(f, sep='\\t', index=False, header=False)\n",
    "\n",
    "chunk_size = 400000\n",
    "num_chunks = len(vcf_df) // chunk_size + (len(vcf_df) % chunk_size > 0)\n",
    "\n",
    "for i in range(num_chunks):\n",
    "    chunk_file = f'./Genebass/SV/sv_unique_chunk_{i+1}.vcf'\n",
    "    start = i * chunk_size\n",
    "    end = start + chunk_size\n",
    "    chunk = vcf_df.iloc[start:end]\n",
    "    \n",
    "    with open(chunk_file, 'w') as f:\n",
    "        f.write(vcf_header)\n",
    "        chunk.to_csv(f, sep='\\t', index=False, header=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "raw",
   "id": "1b7404cb-4987-488c-9591-998cedf36f8a",
   "metadata": {},
   "source": [
    "# Scripts to annotate variants using VEP:\n",
    "\n",
    "ml CPAN\n",
    "ml tabix\n",
    "ml vep/112\n",
    "ml perl/5.40.0\n",
    "ml samtools\n",
    "ml htslib\n",
    "\n",
    "for i in {1..14}; do bsub -q premium -P acc_GENECAD -W 39:59 -n 8 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=4GB] \"vep -i /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}.vcf --fork 4 --cache --dir_cache /hpc/packages/minerva-centos7/vep/112/cache/ --offline --assembly GRCh38 --dir_plugins /hpc/packages/minerva-centos7/vep/112/Plugins/ --output_file /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}_vep.vcf --no_stats --force_overwrite --vcf --symbol --canonical --mane_select --show_ref_allele --plugin dbNSFP,/sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP4.9a_grch38.gz,/sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP_replacement_logic.txt,transcript_match=1,SIFT_pred,SIFT4G_pred,Polyphen2_HDIV_pred,Polyphen2_HVAR_pred,LRT_pred,MutationTaster_pred,MutationAssessor_pred,FATHMM_pred,PROVEAN_pred,MetaSVM_pred,MetaLR_pred,M-CAP_pred,PrimateAI_pred,DEOGEN2_pred,BayesDel_addAF_pred,BayesDel_noAF_pred,ClinPred_pred,LIST-S2_pred,fathmm-MKL_coding_rankscore,fathmm-MKL_coding_pred,fathmm-XF_coding_rankscore,fathmm-XF_coding_pred,Aloft_pred,VEST4_score,VEST4_rankscore,REVEL_score,REVEL_rankscore,MutPred_score,MutPred_rankscore,MVP_score,MVP_rankscore,MPC_score,MPC_rankscore,DANN_score,DANN_rankscore,CADD_raw,CADD_raw_rankscore,Eigen-raw_coding,Eigen-raw_coding_rankscore,Eigen-PC-raw_coding,Eigen-PC-raw_coding_rankscore,AlphaMissense_score,AlphaMissense_rankscore,AlphaMissense_pred,gnomAD_exomes_AF,gnomAD_exomes_POPMAX_AF,gnomAD_genomes_AF,gnomAD_genomes_POPMAX_AF\"; done\n",
    "\n",
    "# Canonical (not used)\n",
    "for i in {1..14}; do bsub -q premium -P acc_GENECAD -W 39:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=3GB] filter_vep -i /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}_vep.vcf -o /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}_vep_canonical.vcf --format vcf --force_overwrite --filter \"(CANONICAL is YES)\" --only_matched; done\n",
    "\n",
    "# Canonical and MANE\n",
    "for i in {1..14}; do bsub -q premium -P acc_GENECAD -W 39:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=3GB] filter_vep -i /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}_vep.vcf -o /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_chunk_${i}_vep_filter.vcf --format vcf --force_overwrite --filter \"(CANONICAL is YES) and (MANE_SELECT exists)\" --only_matched; done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "629e1d1c-e3b8-4bbd-8ca3-fc3500fd3eed",
   "metadata": {},
   "outputs": [],
   "source": [
    "for chunk in range(1,15):\n",
    "    vcf = pd.read_csv(f'./Genebass/SV/sv_unique_chunk_{chunk}_vep_filter.vcf', sep='\\t', skiprows=53)\n",
    "    vcf = vcf.set_axis(['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO'],axis=1).drop(['QUAL','FILTER'],axis=1)\n",
    "    vcf = vcf.assign(INFO=vcf['INFO'].str.split(',')).explode('INFO')\n",
    "    vcf['INFO'] = vcf['INFO'].str.split('|')\n",
    "    \n",
    "    column_names = [\n",
    "        'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature',\n",
    "        'BIOTYPE', 'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
    "        'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'REF_ALLELE',\n",
    "        'DISTANCE', 'STRAND', 'FLAGS', 'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE_SELECT',\n",
    "        'Aloft_pred', 'AlphaMissense_pred', 'AlphaMissense_rankscore', 'AlphaMissense_score',\n",
    "        'BayesDel_addAF_pred', 'BayesDel_noAF_pred', 'CADD_raw', 'CADD_raw_rankscore',\n",
    "        'ClinPred_pred', 'DANN_rankscore', 'DANN_score', 'DEOGEN2_pred', 'Eigen-PC-raw_coding',\n",
    "        'Eigen-PC-raw_coding_rankscore', 'Eigen-raw_coding', 'Eigen-raw_coding_rankscore',\n",
    "        'FATHMM_pred', 'LIST-S2_pred', 'LRT_pred', 'M-CAP_pred', 'MPC_rankscore', 'MPC_score',\n",
    "        'MVP_rankscore', 'MVP_score', 'MetaLR_pred', 'MetaSVM_pred', 'MutPred_rankscore',\n",
    "        'MutPred_score', 'MutationAssessor_pred', 'MutationTaster_pred', 'PROVEAN_pred',\n",
    "        'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_pred', 'PrimateAI_pred', 'REVEL_rankscore',\n",
    "        'REVEL_score', 'SIFT4G_pred', 'SIFT_pred', 'VEST4_rankscore', 'VEST4_score',\n",
    "        'fathmm-MKL_coding_pred', 'fathmm-MKL_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "        'fathmm-XF_coding_rankscore', 'gnomAD_exomes_AF', 'gnomAD_exomes_POPMAX_AF',\n",
    "        'gnomAD_genomes_AF', 'gnomAD_genomes_POPMAX_AF'\n",
    "    ]\n",
    "    \n",
    "    vcf[column_names] = pd.DataFrame(vcf['INFO'].tolist(), index=vcf.index)\n",
    "    vcf = vcf.drop(columns=['INFO'])\n",
    "    vcf = vcf.drop(['Allele','Feature_type','Feature','BIOTYPE','EXON','INTRON','HGVSc','HGVSp',\n",
    "                    'cDNA_position','CDS_position','Protein_position','Amino_acids','Codons','Existing_variation',\n",
    "                    'REF_ALLELE','DISTANCE','STRAND','FLAGS','SYMBOL_SOURCE','HGNC_ID',\n",
    "                    'gnomAD_genomes_AF','gnomAD_genomes_POPMAX_AF'],axis=1)\n",
    "    \n",
    "    vcf.loc[vcf['Consequence'].str.contains('stop_lost|start_lost|transcript_amplification|inframe_insertion|inframe_deletion|missense_variant|protein_altering_variant',na=False),'snp_type'] = 'Missense'\n",
    "    vcf.loc[vcf['Consequence'].str.contains('transcript_ablation|splice_acceptor_variant|splice_donor_variant|stop_gained|frameshift_variant',na=False),'snp_type'] = 'PTV'\n",
    "    vcf = vcf.loc[vcf['IMPACT'] != 'MODIFIER']\n",
    "    \n",
    "    #####\n",
    "    \n",
    "    qualitative_tools = ['SIFT_pred', 'SIFT4G_pred', 'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_pred', \n",
    "                         'LRT_pred', 'MutationTaster_pred', 'FATHMM_pred', 'PROVEAN_pred', \n",
    "                         'MetaSVM_pred', 'MetaLR_pred', 'M-CAP_pred', 'PrimateAI_pred', \n",
    "                         'DEOGEN2_pred', 'BayesDel_addAF_pred', 'BayesDel_noAF_pred', \n",
    "                         'ClinPred_pred', 'LIST-S2_pred', 'fathmm-MKL_coding_pred', \n",
    "                         'fathmm-XF_coding_pred', 'MutationAssessor_pred', 'Aloft_pred']\n",
    "    \n",
    "    quantitative_tools = ['VEST4_rankscore', 'REVEL_rankscore', 'MutPred_rankscore', \n",
    "                          'MVP_rankscore', 'MPC_rankscore', 'DANN_rankscore', 'CADD_raw_rankscore', \n",
    "                          'Eigen-raw_coding_rankscore', 'Eigen-PC-raw_coding_rankscore']\n",
    "    \n",
    "    \n",
    "    def calculate_score(row):\n",
    "        score = 0\n",
    "        count = 0\n",
    "        \n",
    "        # Qualitative scoring\n",
    "        for tool in qualitative_tools:\n",
    "            if tool in row and row[tool] != '':\n",
    "                if tool == 'MutationAssessor_pred' and row[tool] == 'H':\n",
    "                    score += 1\n",
    "                elif tool == 'Aloft_pred' and row[tool] in ['R', 'D']:\n",
    "                    score += 1\n",
    "                elif tool == 'MutationTaster_pred' and 'D' in str(row[tool]):\n",
    "                    score += 1\n",
    "                elif row[tool] == 'D':\n",
    "                    score += 1\n",
    "                count += 1\n",
    "        \n",
    "        # Quantitative scoring\n",
    "        for tool in quantitative_tools:\n",
    "            if tool in row and row[tool] != '':\n",
    "                try:\n",
    "                    if float(row[tool]) > 0.9:\n",
    "                        score += 1\n",
    "                    count += 1\n",
    "                except ValueError:\n",
    "                    pass  # Ignore non-numeric values\n",
    "    \n",
    "        return score, count\n",
    "    \n",
    "    # Apply the scoring function\n",
    "    vcf[['total_score', 'non_missing_count']] = vcf.apply(lambda row: pd.Series(calculate_score(row)), axis=1)\n",
    "    \n",
    "    vcf['Pred5'] = 0\n",
    "    vcf.loc[vcf['SIFT_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "    vcf.loc[vcf['LRT_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "    vcf.loc[vcf['MutationTaster_pred'].str.contains('A|D',na=False), 'Pred5'] += 1\n",
    "    vcf.loc[vcf['Polyphen2_HDIV_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "    vcf.loc[vcf['Polyphen2_HVAR_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "    \n",
    "    #####\n",
    "    \n",
    "    vcf = vcf[['ID','CHROM','POS','REF','ALT','snp_type','Consequence','IMPACT','SYMBOL',\n",
    "               'total_score','non_missing_count','Pred5',\n",
    "               'AlphaMissense_pred','AlphaMissense_rankscore','AlphaMissense_score',\n",
    "               'REVEL_rankscore','REVEL_score',\n",
    "               'gnomAD_exomes_AF','gnomAD_exomes_POPMAX_AF']]\n",
    "    vcf.to_csv(f'./Genebass/SV/sv_unique_chunk_{chunk}_vep_filter_cleaned.vcf', sep='\\t', index=None)\n",
    "\n",
    "# Combine chunk outputs\n",
    "file_paths = [f\"./Genebass/SV/sv_unique_chunk_{chunk}_vep_filter_cleaned.vcf\" for chunk in range(1, 15)]\n",
    "dfs = [pd.read_csv(file, sep='\\t', comment='#') for file in file_paths]\n",
    "final_df = pd.concat(dfs, ignore_index=True)\n",
    "final_df.to_pickle('./Genebass/SV/vep_combined.pkl')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b29e6c4a-bc40-41e1-9a8f-856b4bc5b713",
   "metadata": {},
   "source": [
    "#### Running LOFTEE on PTVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7364df4-64e0-4fff-93e9-1242615cc8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gbv = pd.read_pickle('./Genebass/SV/vep_combined.pkl')\n",
    "\n",
    "df = gbv.loc[gbv['snp_type'] == 'PTV']\n",
    "df = df[['CHROM','POS','ID','REF','ALT']].set_axis(['chr','pos','id','ref','alt'],axis=1)\n",
    "df['qual'] = '.'\n",
    "df['filter'] = '.'\n",
    "df['info'] = '.'\n",
    "\n",
    "# Reorder columns to VCF format\n",
    "vcf_columns = ['chr', 'pos', 'id', 'ref', 'alt', 'qual', 'filter', 'info']\n",
    "vcf_df = df[vcf_columns]\n",
    "vcf_df = vcf_df.sort_values(['chr','pos','ref','alt'])\n",
    "\n",
    "# Save as VCF\n",
    "vcf_header = \"\"\"##fileformat=VCFv4.2\n",
    "##source=CustomGeneratedVCF\n",
    "#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\n",
    "\"\"\"\n",
    "with open('./Genebass/SV/sv_unique_lof.vcf', 'w') as f:\n",
    "    f.write(vcf_header)\n",
    "    vcf_df.to_csv(f, sep='\\t', index=False, header=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "raw",
   "id": "58c45c2a-83ff-40cd-954d-1d4f812168ad",
   "metadata": {},
   "source": [
    "# Scripts to annotate variants using LOFTEE:\n",
    "\n",
    "ml CPAN\n",
    "ml tabix\n",
    "ml vep/112\n",
    "ml perl/5.40.0\n",
    "ml samtools\n",
    "ml htslib\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 8 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=4GB] \"vep -i /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_lof.vcf --fork 4 --cache --dir_cache /hpc/packages/minerva-centos7/vep/112/cache/ --offline --assembly GRCh38 --dir_plugins /sc/arion/projects/GENECAD/Robert/Software/loftee-1.0.4/ --output_file /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_lof_loftee.vcf --no_stats --force_overwrite --vcf --symbol --canonical --mane_select --show_ref_allele --plugin LoF,loftee_path:/sc/arion/projects/GENECAD/Robert/Software/loftee-1.0.4/,human_ancestor_fa:/sc/arion/projects/GENECAD/Robert/Software/human_ancestor/human_ancestor.fa.gz,conservation_file:/sc/arion/projects/GENECAD/Robert/Software/conservation_file/loftee.sql\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=3GB] filter_vep -i /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_lof_loftee.vcf -o /sc/arion/projects/GENECAD/Robert/DOE/Genebass/SV/sv_unique_lof_loftee_filter.vcf --format vcf --force_overwrite --filter \"(CANONICAL is YES) and (MANE_SELECT exists)\" --only_matched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f3b12a7-30d5-490f-86d7-17b6e8ebae42",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = pd.read_csv(f'./Genebass/SV/sv_unique_lof_loftee_filter.vcf', sep='\\t', skiprows=53)\n",
    "vcf = vcf.set_axis(['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO'],axis=1).drop(['QUAL','FILTER'],axis=1)\n",
    "vcf = vcf.assign(INFO=vcf['INFO'].str.split(',')).explode('INFO')\n",
    "vcf['INFO'] = vcf['INFO'].str.split('|')\n",
    "\n",
    "column_names = [\n",
    "    'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature',\n",
    "    'BIOTYPE', 'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
    "    'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'REF_ALLELE',\n",
    "    'DISTANCE', 'STRAND', 'FLAGS', 'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE_SELECT',\n",
    "    'LoF','LoF_filter','LoF_flags','LoF_info'\n",
    "]\n",
    "\n",
    "vcf[column_names] = pd.DataFrame(vcf['INFO'].tolist(), index=vcf.index)\n",
    "vcf = vcf.drop(columns=['INFO'])\n",
    "\n",
    "vcf.loc[vcf['Consequence'].str.contains('stop_lost|start_lost|transcript_amplification|inframe_insertion|inframe_deletion|missense_variant|protein_altering_variant',na=False),'snp_type'] = 'Missense'\n",
    "vcf.loc[vcf['Consequence'].str.contains('transcript_ablation|splice_acceptor_variant|splice_donor_variant|stop_gained|frameshift_variant',na=False),'snp_type'] = 'PTV'\n",
    "vcf = vcf.loc[vcf['IMPACT'] != 'MODIFIER']\n",
    "vcf = vcf[['ID','Consequence','SYMBOL','LoF','LoF_filter','LoF_flags','LoF_info']]\n",
    "vcf.to_pickle('./Genebass/SV/vep_loftee.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86cb9bf2-5cf2-4616-8947-e4735ffdb020",
   "metadata": {},
   "source": [
    "#### Combining VEP files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ddfd101-390f-4758-a276-4a656840a34e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vep = pd.read_pickle('./Genebass/SV/vep_combined.pkl')\n",
    "loftee = pd.read_pickle('./Genebass/SV/vep_loftee.pkl')\n",
    "vep = vep.merge(loftee, how='left')\n",
    "vep['missense_score'] = vep['total_score']/vep['non_missing_count']\n",
    "vep = vep.drop(['LoF_filter','LoF_info'],axis=1)\n",
    "\n",
    "gof = pd.read_pickle('./LoGoFunc/all_predictions.pkl')\n",
    "gof = gof.rename({'#CHROM':'CHROM'},axis=1).drop(['ID','prediction','LoGoFunc_LOF','SYMBOL'],axis=1)\n",
    "vep = vep.merge(gof, how='left')\n",
    "\n",
    "lof = pd.read_pickle('./LoGoFunc/LOF_predictions.pkl').rename({'#CHROM':'CHROM'},axis=1).drop(['ID','SYMBOL'],axis=1)\n",
    "vep = vep.merge(lof, how='left')\n",
    "\n",
    "vep = vep[['ID','CHROM','POS','REF','ALT','snp_type','Consequence','SYMBOL',\n",
    "           'non_missing_count','missense_score','LoF','LoGoFunc_GOF','LoGoFunc_LOF']]\n",
    "\n",
    "vep['annotation'] = '8_other_missense'\n",
    "vep.loc[(vep['non_missing_count'] > 7) & (vep['missense_score'] > 0.5), 'annotation'] = '7_other_missense_0.5'\n",
    "vep.loc[vep['LoGoFunc_GOF'] > 0, 'annotation'] = '6_gof_missense'\n",
    "vep.loc[(vep['LoGoFunc_GOF'] > 0) & (vep['non_missing_count'] > 7) & (vep['missense_score'] > 0.5), 'annotation'] = '5_gof_missense_0.5'\n",
    "vep.loc[vep['LoGoFunc_LOF'] > 0, 'annotation'] = '4_lof_missense'\n",
    "vep.loc[(vep['LoGoFunc_LOF'] > 0) & (vep['non_missing_count'] > 7) & (vep['missense_score'] > 0.5), 'annotation'] = '3_lof_missense_0.5'\n",
    "vep.loc[(vep['snp_type'] == 'PTV') | (vep['LoF'] == 'LC'), 'annotation'] = '2_lof_lc'\n",
    "vep.loc[(vep['LoF'] == 'HC'), 'annotation'] = '1_lof_hc'\n",
    "vep = vep.sort_values('annotation').drop_duplicates('ID')\n",
    "\n",
    "vep.loc[vep['LoGoFunc_GOF'] >= 2/3, 'hc_gof'] = 1\n",
    "\n",
    "vep.to_pickle('./Genebass/SV/vep_cleaned.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab722dc2-8a76-48f0-a24a-78a291344390",
   "metadata": {},
   "source": [
    "## Jurgens et al. rare variant analysis"
   ]
  },
  {
   "cell_type": "raw",
   "id": "40777581-1ae2-43bc-8ce1-b57d245b07b0",
   "metadata": {},
   "source": [
    "# Download data\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies && wget -P /sc/arion/projects/GENECAD/Robert/DOE/Jurgens https://api.kpndataregistry.org/api/d/4qcL75\"\n",
    "\n",
    "# Extract relevant columns\n",
    "awk -F'\\t' 'BEGIN {\n",
    "    OFS=\"\\t\"\n",
    "    # List of desired columns in the order you want them\n",
    "    desired_cols = \"gene\\tphenotype\\tEst_Meta__hclof_noflag_POPMAX0.001\\tP_Meta_overlapcorrected__hclof_noflag_POPMAX0.001\\tEst_Meta__hclof_noflag_missense0.8_POPMAX0.001\\tP_Meta_overlapcorrected__hclof_noflag_missense0.8_POPMAX0.001\\tEst_Meta__hclof_noflag_missense0.5_POPMAX0.001\\tP_Meta_overlapcorrected__hclof_noflag_missense0.5_POPMAX0.001\\tEst_Meta__hclof_noflag_missense0.5_POPMAX0.00001\\tP_Meta_overlapcorrected__hclof_noflag_missense0.5_POPMAX0.00001\\tEst_Meta__missense0.5_POPMAX0.00001\\tP_Meta_overlapcorrected__missense0.5_POPMAX0.00001\\tEst_Meta__missense0.2_POPMAX0.00001\\tP_Meta_overlapcorrected__missense0.2_POPMAX0.00001\\tEst_Meta__hclof_noflag_POPMAX0.01\\tP_Meta_overlapcorrected__hclof_noflag_POPMAX0.01\\tEst_Meta__hclof_noflag_missense0.8_POPMAX0.01\\tP_Meta_overlapcorrected__hclof_noflag_missense0.8_POPMAX0.01\\tEst_Meta__hclof_noflag_missense0.5_POPMAX0.01\\tP_Meta_overlapcorrected__hclof_noflag_missense0.5_POPMAX0.01\"\n",
    "    split(desired_cols, colList, \"\\t\")\n",
    "}\n",
    "\n",
    "NR==1 {\n",
    "    # Map column names to their indices\n",
    "    for (i=1; i<=NF; i++) {\n",
    "        colIdx[$i] = i\n",
    "    }\n",
    "    # Print the header line in desired order\n",
    "    for (j=1; j<=length(colList); j++) {\n",
    "        printf \"%s%s\", colList[j], (j<length(colList)?OFS:ORS)\n",
    "    }\n",
    "    next\n",
    "}\n",
    "\n",
    "{\n",
    "    # Print the data columns in the specified order\n",
    "    for (j=1; j<=length(colList); j++) {\n",
    "        printf \"%s%s\", $colIdx[colList[j]], (j<length(colList)?OFS:ORS)\n",
    "    }\n",
    "}' /sc/arion/projects/GENECAD/Robert/DOE/Jurgens/UKB_AoU_MGB_overlapcorrected.tsv > /sc/arion/projects/GENECAD/Robert/DOE/Jurgens/UKB_AoU_MGB_overlapcorrected_cleaned.tsv\n",
    "\n",
    "# Input cleaned file (adjust path if needed)\n",
    "INPUT=\"/sc/arion/projects/GENECAD/Robert/DOE/Jurgens/UKB_AoU_MGB_overlapcorrected_cleaned.tsv\"\n",
    "OUTDIR=\"$(dirname \"$INPUT\")\"\n",
    "\n",
    "# Read the header line\n",
    "header=$(head -n 1 \"$INPUT\")\n",
    "\n",
    "# Convert header to array of column names\n",
    "IFS=$'\\t' read -r -a columns <<< \"$header\"\n",
    "\n",
    "# Identify indices for gene and phenotype\n",
    "gene_idx=0\n",
    "phenotype_idx=0\n",
    "\n",
    "declare -A mask_pairs\n",
    "\n",
    "# Find gene and phenotype columns\n",
    "for i in \"${!columns[@]}\"; do\n",
    "    if [ \"${columns[$i]}\" == \"gene\" ]; then\n",
    "        gene_idx=$((i+1))\n",
    "    elif [ \"${columns[$i]}\" == \"phenotype\" ]; then\n",
    "        phenotype_idx=$((i+1))\n",
    "    fi\n",
    "done\n",
    "\n",
    "# Identify all masks by looking for columns starting with Est_Meta__ and P_Meta_overlapcorrected__\n",
    "# The masks should be the substring following Est_Meta__ or P_Meta_overlapcorrected__ respectively.\n",
    "for i in \"${!columns[@]}\"; do\n",
    "    col=\"${columns[$i]}\"\n",
    "    # Check if column name starts with Est_Meta__\n",
    "    if [[ \"$col\" == Est_Meta__* ]]; then\n",
    "        # Extract the mask name by removing the prefix 'Est_Meta__'\n",
    "        mask=\"${col#Est_Meta__}\"\n",
    "        # Store the Est_Meta column index\n",
    "        mask_pairs[\"$mask,Est\"]=$((i+1))\n",
    "    fi\n",
    "    # Check if column name starts with P_Meta_overlapcorrected__\n",
    "    if [[ \"$col\" == P_Meta_overlapcorrected__* ]]; then\n",
    "        # Extract the mask name by removing prefix 'P_Meta_overlapcorrected__'\n",
    "        mask=\"${col#P_Meta_overlapcorrected__}\"\n",
    "        # Store the P_Meta column index\n",
    "        mask_pairs[\"$mask,P\"]=$((i+1))\n",
    "    fi\n",
    "done\n",
    "\n",
    "# Now we have pairs of columns for each mask: Est and P.\n",
    "# Iterate over the unique masks and create a separate file for each.\n",
    "declare -A unique_masks\n",
    "for key in \"${!mask_pairs[@]}\"; do\n",
    "    # key format: \"mask,Est\" or \"mask,P\"\n",
    "    mask=\"${key%,*}\"\n",
    "    unique_masks[\"$mask\"]=1\n",
    "done\n",
    "\n",
    "for mask in \"${!unique_masks[@]}\"; do\n",
    "    est_col_idx=\"${mask_pairs[\"$mask,Est\"]}\"\n",
    "    p_col_idx=\"${mask_pairs[\"$mask,P\"]}\"\n",
    "    \n",
    "    # Output file name\n",
    "    outfile=\"${OUTDIR}/UKB_AoU_MGB_overlapcorrected_cleaned_${mask}.tsv\"\n",
    "    \n",
    "    # Print header\n",
    "    echo -e \"gene\\tphenotype\\tEst_Meta__${mask}\\tP_Meta_overlapcorrected__${mask}\" > \"$outfile\"\n",
    "    \n",
    "    # Print data lines\n",
    "    awk -F'\\t' -v OFS='\\t' \\\n",
    "        -v g=\"$gene_idx\" \\\n",
    "        -v p=\"$phenotype_idx\" \\\n",
    "        -v e=\"$est_col_idx\" \\\n",
    "        -v pm=\"$p_col_idx\" 'NR>1 {print $g, $p, $e, $pm}' \"$INPUT\" >> \"$outfile\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e48f8e3a-4747-4ea8-a07e-36bef5447ee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.DataFrame()\n",
    "\n",
    "masks = ['hclof_noflag_missense0.5_POPMAX0.00001','hclof_noflag_missense0.5_POPMAX0.01',\n",
    "         'hclof_noflag_POPMAX0.01','missense0.5_POPMAX0.00001',\n",
    "         'hclof_noflag_missense0.8_POPMAX0.001','missense0.2_POPMAX0.00001',\n",
    "         'hclof_noflag_missense0.8_POPMAX0.01','hclof_noflag_POPMAX0.001',\n",
    "         'hclof_noflag_missense0.5_POPMAX0.001']\n",
    "\n",
    "for mask in masks:\n",
    "    ss = pd.read_csv(f'./Jurgens/UKB_AoU_MGB_overlapcorrected_cleaned_{mask}.tsv', sep='\\t').dropna().set_axis(['gene','phenotype','beta','p'],axis=1)\n",
    "    ss = ss.loc[ss['p'] < 0.05]\n",
    "    ss['p'] = -np.log10(ss['p'])\n",
    "    ss.loc[ss['p'] > 300, 'p'] = 300\n",
    "    ss['mask'] = mask\n",
    "    results = pd.concat([results,ss])\n",
    "    \n",
    "results = results.rename({'gene':'id'},axis=1)\n",
    "targets = pd.read_pickle('./OT/Raw/targets.pkl')\n",
    "targets = targets[['id','approvedSymbol']].rename({'approvedSymbol':'gene'},axis=1)\n",
    "results = results.merge(targets).drop('id',axis=1)\n",
    "results['mask'] = results['mask'].str.replace('hclof_noflag_','lof_').str.replace('missense','ms').str.replace('POPMAX','af')\n",
    "results.to_pickle('./Jurgens/UKB_AoU_MGB_combined.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c151f9-c0b3-4e76-a3b4-6a3e3ae015a3",
   "metadata": {},
   "source": [
    "## PanUKBB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4467fd77-1a49-40aa-b824-f483e0eb1df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pu = pd.read_excel('./PanUKBB/panukbb_pheno.xlsx', usecols = ['trait_type','phenocode','description','aws_link'])\n",
    "\n",
    "pu_icd = pu.loc[pu['trait_type'] == 'icd10']\n",
    "pu_icd = pu_icd.loc[(pu_icd['phenocode'].isin(icd['ICD'])) | (pu_icd['phenocode'].isin(custom['icd']))]\n",
    "\n",
    "pu_p12 = pu.loc[pu['trait_type'] == 'phecode']\n",
    "pu_p12['phenocode'] = pu_p12['phenocode'].astype(float)\n",
    "pu_p12 = pu_p12.loc[(pu_p12['phenocode'].isin(pu_p12['Phecode'])) | (pu_p12['phenocode'].isin(custom['phecode1.2']))]\n",
    "\n",
    "pu = pd.concat([pu_icd,pu_p12])\n",
    "\n",
    "for i, chunk in enumerate(range(0, len(pu), 100)):\n",
    "    pu['aws_link'].iloc[chunk:chunk+100].to_csv(f'/sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Download/download_links_{i+1}.txt', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "72cfe8c7-84d2-4324-bcc8-9c57376f4869",
   "metadata": {},
   "source": [
    "# Download data and extract variants with p < 0.05\n",
    "for i in {2..11}; do\n",
    "    bsub -q premium -P acc_GENECAD -W 29:59 -n 4 -o run.out -e run.error \\\n",
    "    -R span[hosts=1] -R rusage[mem=2GB] \\\n",
    "    \"ml proxies aria2 pigz && \\\n",
    "    while read -r url; do\n",
    "        filename=\\$(basename \\\"\\$url\\\");\n",
    "        sig_filename=\\\"\\${filename%.gz}.sig\\\";\n",
    "        sig_filepath=\\\"/sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Processed/\\$sig_filename\\\";\n",
    "        if ! aria2c \\\"\\$url\\\"; then\n",
    "            continue;\n",
    "        fi;\n",
    "        # Check if the file contains 'neglog10_pval_EUR'\n",
    "        if ! pigz -dc \\\"\\$filename\\\" | head -n1 | grep -q \\\"neglog10_pval_EUR\\\"; then\n",
    "            rm \\\"\\$filename\\\";\n",
    "            continue;\n",
    "        fi;\n",
    "        if pigz -dc \\\"\\$filename\\\" | \\\n",
    "        awk -F'\\t' '\n",
    "            BEGIN {col=0}\n",
    "            NR==1{\n",
    "                for(i=1;i<=NF;i++){\n",
    "                    if(\\$i==\\\"neglog10_pval_EUR\\\"){\n",
    "                        col=i;\n",
    "                        break;\n",
    "                    }\n",
    "                }\n",
    "                if(col==0){\n",
    "                    exit 1;  # Column not found; exit with error\n",
    "                }\n",
    "                print;  # Output the header row\n",
    "            }\n",
    "            NR>1 && \\$col > 1.30102999 && \\$col != \\\"NA\\\" {\n",
    "                print;\n",
    "            }\n",
    "        ' > \\\"\\$sig_filepath\\\"; then\n",
    "            rm \\\"\\$filename\\\";\n",
    "        else\n",
    "            rm \\\"\\$filename\\\";  # Ensure the file is deleted in case of processing failure\n",
    "            continue;\n",
    "        fi;\n",
    "    done < /sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Download/download_links_${i}.txt\";\n",
    "done\n",
    "\n",
    "# Clean files\n",
    "bsub -q premium -P acc_GENECAD -W 29:59 -n 4 -o run.out -e run.error -R \"span[hosts=1]\" -R \"rusage[mem=2GB]\" <<'EOF'\n",
    "indir=\"/sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Processed\"\n",
    "outdir=\"/sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Cleaned\"\n",
    "for f in \"$indir\"/*.tsv.bgz.sig; do\n",
    "header=$(head -1 \"$f\")\n",
    "if echo \"$header\" | grep -q \"beta_meta\"; then\n",
    "cols=\"1,2,3,4,$(awk -F'\\t' '{for(i=1;i<=NF;i++){if($i==\"beta_meta\")b=i;if($i==\"se_meta\")s=i;if($i==\"neglog10_pval_meta\")p=i}}END{print b\",\"s\",\"p}' <(echo \"$header\"))\"\n",
    "else\n",
    "cols=\"1,2,3,4,$(awk -F'\\t' '{for(i=1;i<=NF;i++){if($i==\"beta_EUR\")b=i;if($i==\"se_EUR\")s=i;if($i==\"neglog10_pval_EUR\")p=i}}END{print b\",\"s\",\"p}' <(echo \"$header\"))\"\n",
    "fi\n",
    "awk -F'\\t' -v OFS='\\t' -v c=\"$cols\" 'NR==1{print \"chr\",\"pos\",\"ref\",\"alt\",\"beta\",\"se\",\"neglog10_pval\";next}{split(c,a,\",\");print $a[1],$a[2],$a[3],$a[4],$a[5],$a[6],$a[7]}' \"$f\" > \"$outdir\"/$(basename \"$f\" .tsv.bgz.sig).tsv\n",
    "done\n",
    "EOF\n",
    "\n",
    "# Subset files for eQTL analysis (see eQTL.ipynb)\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" \\\n",
    "'for f in /sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Cleaned/*.tsv; do awk -F\"\\t\" '\\''NR==FNR{a[$1];next}FNR==1{OFS=\"\\t\";print \"id\",\"beta\",\"se\",\"neglog10_pval\"}NR>1{id=$1\"_\"$2\"_\"$3\"_\"$4;if(id in a)print id,$5,$6,$7}'\\'' /sc/arion/projects/GENECAD/Robert/DOE/eQTL/panukbb_unique_variants.tsv \"$f\" > \"/sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/Subset/$(basename \"$f\")\"; done'\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3de17248-5282-4881-a105-75a4c07ac823",
   "metadata": {},
   "source": [
    "# Get variant QC metrics\n",
    "wget https://pan-ukb-us-east-1.s3.amazonaws.com/sumstats_release/full_variant_qc_metrics.txt.bgz\n",
    "\n",
    "input_file=\"full_variant_qc_metrics.txt\"\n",
    "output_file=\"full_variant_qc_metrics_filtered.txt\"\n",
    "\n",
    "awk -F'\\t' '\n",
    "    NR==1 {\n",
    "        # Print header columns 1,2,3,4,5,6,9,11,34,37\n",
    "        print $1,$2,$3,$4,$5,$6,$9,$11,$34,$37\n",
    "        next\n",
    "    }\n",
    "    ($11 >= 0.8 && $34 >= 0.001) {\n",
    "        # Print the same columns for rows passing the filter\n",
    "        print $1,$2,$3,$4,$5,$6,$9,$11,$34,$37\n",
    "    }\n",
    "' OFS='\\t' \"$input_file\" > \"$output_file\"\n",
    "\n",
    "input_file=\"full_variant_qc_metrics_filtered.txt\"\n",
    "output_file=\"full_variant_qc_metrics_filtered.bim\"\n",
    "\n",
    "awk -F'\\t' 'NR==1 { next } {\n",
    "  # Columns:\n",
    "  # 1: chrom, 2: pos, 3: ref, 4: alt, 5: rsid, 6: varid\n",
    "  # BIM format: chrom varid genetic_distance pos allele1 allele2\n",
    "  print $1, $6, 0, $2, $3, $4\n",
    "}' OFS='\\t' \"$input_file\" > \"$output_file\"\n",
    "\n",
    "# Convert to Hg38\n",
    "perl /sc/arion/projects/GENECAD/Robert/tlo/tripleliftover_v133.pl --bim /sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/full_variant_qc_metrics_filtered.bim --base hg19 --target Hg38 --outprefix /sc/arion/projects/GENECAD/Robert/DOE/PanUKBB/full_variant_qc_metrics_filtered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bc97030-c33e-4ab8-b8de-0bfc8880d7da",
   "metadata": {},
   "outputs": [],
   "source": [
    "pu_var = pd.read_csv('./PanUKBB/full_variant_qc_metrics_filtered.bim', sep='\\t', header=None)[[1,0,3,4,5]].set_axis(['id','chr','pos','ref','alt'],axis=1)\n",
    "pu_conv = pd.read_csv('./PanUKBB/full_variant_qc_metrics_filtered.invr.txt', sep='\\t')\n",
    "pu_conv = pu_conv.drop('category',axis=1).set_axis(['id','chr','pos_38','pos'],axis=1)\n",
    "pu_var = pu_var.merge(pu_conv)\n",
    "pu_var['id_19'] = pu_var['chr'].astype(str) + '_' + pu_var['pos'].astype(str) + '_' + pu_var['ref'] + '_' + pu_var['alt']\n",
    "pu_var['id_38'] = pu_var['chr'].astype(str) + '_' + pu_var['pos_38'].astype(str) + '_' + pu_var['ref'] + '_' + pu_var['alt']\n",
    "pu_var[['id_19','id_38','chr','pos','ref','alt']].to_pickle('./PanUKBB/variant_conv.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb7336dc-cb9c-4794-bfb2-0c4f3f81f0c2",
   "metadata": {},
   "source": [
    "## Finngen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affc8e1c-9538-4c95-9d3a-bb2cada203b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find relevant phenotypes\n",
    "fgm = pd.read_csv('./Finngen/finngen_R12_manifest.tsv', sep='\\t')\n",
    "fg = pd.read_excel('./Finngen/finngen_R12_endpoints.xlsx')[['LEVEL','NAME','LONGNAME','HD_ICD_10','COD_ICD_10','version','PARENT']]\n",
    "fg = fg.loc[fg['NAME'].isin(fgm['phenocode'])]\n",
    "fg['ICD'] = fg['HD_ICD_10'].str.replace('.','')\n",
    "\n",
    "a = fg.loc[fg['ICD'].isin(icd['ICD'])]\n",
    "b = fg.loc[fg['LONGNAME'].isin(custom['finngen'])]\n",
    "c = fg.loc[(~fg['ICD'].isin(icd['ICD'])) & (fg['HD_ICD_10'].notna())]\n",
    "c['ICD_3c'] = c['ICD'].str[:3]\n",
    "c = c.loc[c['ICD_3c'].isin(icd['ICD'])]\n",
    "c_inc = pd.read_excel('./Finngen/bar_pheno.xlsx')\n",
    "c_inc = c_inc.loc[c_inc['Include'] == 'Y']['NAME']\n",
    "c = c.loc[(~c['HD_ICD_10'].str.contains('|', regex=False)) | (c['NAME'].isin(c_inc))]\n",
    "\n",
    "fg = pd.concat([a,b,c])\n",
    "fgm = fgm.loc[fgm['phenocode'].isin(fg['NAME'])]\n",
    "fgm['path_https'].to_csv('./Finngen/download_links.txt', index=False, header=False)\n",
    "\n",
    "for i, chunk in enumerate(range(0, len(fgm), 100)):\n",
    "    fgm['path_https'].iloc[chunk:chunk+100].to_csv(f'/sc/arion/projects/GENECAD/Robert/DOE/Finngen/Download/download_links_{i+1}.txt', index=False, header=False)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8bc23242-e6f8-4781-85f2-7e4a1dfe92a9",
   "metadata": {},
   "source": [
    "# Download and extract data\n",
    "for i in {1..15}; do bsub -q premium -P acc_GENECAD -W 29:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=2GB] \"ml proxies aria2 pigz && while read -r url; do filename=\\$(basename \"\\$url\"); sig_filename=\"\\${filename%.gz}.sig\"; sig_filepath=\"/sc/arion/projects/GENECAD/Robert/DOE/Finngen/Processed/\\$sig_filename\"; if ! aria2c \"\\$url\"; then continue; fi; if pigz -dc \"\\$filename\" | awk -F'\\t' '\\$7 < 0.05' > \"\\$sig_filepath\"; then rm \"\\$filename\"; else continue; fi; done < /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Download/download_links_${i}.txt\"; done\n",
    "\n",
    "# Download and extract variant annotations\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"ml proxies axel pigz && wget -P /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations https://storage.googleapis.com/finngen-public-data-r12/annotations/finnge_R12_annotated_variants_v1.gz\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=1GB] \"gunzip /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/finnge_R12_annotated_variants_v1.gz\"\n",
    "\n",
    "# Clean files\n",
    "bsub -q premium -P acc_GENECAD -W 29:59 -n 4 -o run.out -e run.error -R \"span[hosts=1]\" -R \"rusage[mem=2GB]\" \\\n",
    "\"awk -F'\\t' 'NR==1 {\n",
    "  for (i=1; i<=NF; i++) {col[\\$i]=i}\n",
    "  # Print header line for selected columns\n",
    "  print \\\"#variant\\\",\\\"chr\\\",\\\"pos\\\",\\\"ref\\\",\\\"alt\\\",\\\"INFO\\\",\\\"gene_most_severe\\\",\\\"most_severe\\\",\\\"rsid\\\",\\\"b37_coord\\\",\\\"EXOME_AF_fin\\\",\\\"EXOME_AF_nfe\\\",\\\"GENOME_AF_fin\\\",\\\"GENOME_AF_nfe\\\"\n",
    "} \n",
    "NR>1 {\n",
    "  print \\\n",
    "    \\$col[\\\"#variant\\\"],\\\n",
    "    \\$col[\\\"chr\\\"],\\\n",
    "    \\$col[\\\"pos\\\"],\\\n",
    "    \\$col[\\\"ref\\\"],\\\n",
    "    \\$col[\\\"alt\\\"],\\\n",
    "    \\$col[\\\"INFO\\\"],\\\n",
    "    \\$col[\\\"gene_most_severe\\\"],\\\n",
    "    \\$col[\\\"most_severe\\\"],\\\n",
    "    \\$col[\\\"rsid\\\"],\\\n",
    "    \\$col[\\\"b37_coord\\\"],\\\n",
    "    \\$col[\\\"EXOME_AF_fin\\\"],\\\n",
    "    \\$col[\\\"EXOME_AF_nfe\\\"],\\\n",
    "    \\$col[\\\"GENOME_AF_fin\\\"],\\\n",
    "    \\$col[\\\"GENOME_AF_nfe\\\"]\n",
    "}' OFS='\\t' /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/var_anno.txt > /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/var_anno_filt.txt\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1beffb-4eab-4a0b-b2b9-2b3b5b065877",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./Finngen/Annotations/var_anno_filt.txt', sep='\\t')\n",
    "df = df.loc[df['INFO'] >= 0.8]\n",
    "df['max_af'] = df[['EXOME_AF_fin','GENOME_AF_fin']].max(axis=1)\n",
    "df.loc[df['most_severe'].str.contains('stop_lost|start_lost|transcript_amplification|inframe_insertion|inframe_deletion|missense_variant|protein_altering_variant',na=False),'snp_type'] = 'Missense'\n",
    "df.loc[df['most_severe'].str.contains('transcript_ablation|splice_acceptor_variant|splice_donor_variant|stop_gained|frameshift_variant',na=False),'snp_type'] = 'PTV'\n",
    "\n",
    "# Save common variants\n",
    "cl = df.loc[df['max_af'] >= 0.01]\n",
    "cl = cl[['#variant','chr','pos','ref','alt']].rename({'#variant':'id'},axis=1)\n",
    "cl.to_csv('./Finngen/Annotations/common_var.tsv', sep='\\t', index=False)\n",
    "\n",
    "# Save rare variants as VCF\n",
    "ml = df.loc[(df['max_af'] < 0.01) & (df['snp_type'].notna())]\n",
    "ml = ml[['chr','pos','#variant','ref','alt']].rename({'#variant':'id'},axis=1)\n",
    "ml.loc[ml['chr'] == 23, 'chr'] = 'X'\n",
    "\n",
    "ml['qual'] = '.'\n",
    "ml['filter'] = '.'\n",
    "ml['info'] = '.'\n",
    "\n",
    "vcf_columns = ['chr', 'pos', 'id', 'ref', 'alt', 'qual', 'filter', 'info']\n",
    "vcf_df = ml[vcf_columns]\n",
    "vcf_df = vcf_df.sort_values(['chr','pos','ref','alt'])\n",
    "\n",
    "vcf_header = \"\"\"##fileformat=VCFv4.2\n",
    "##source=CustomGeneratedVCF\n",
    "#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\n",
    "\"\"\"\n",
    "with open('./Finngen/Annotations/sv_unique.vcf', 'w') as f:\n",
    "    f.write(vcf_header)\n",
    "    vcf_df.to_csv(f, sep='\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b33626b-9097-4a9d-989d-e6a0010bce00",
   "metadata": {},
   "source": [
    "### Run VEP for rare missense and LOF variants"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e0bee32f-28c9-4ef4-974d-445d96ae3121",
   "metadata": {},
   "source": [
    "# Annotate using VEP:\n",
    "ml CPAN\n",
    "ml tabix\n",
    "ml vep/112\n",
    "ml perl/5.40.0\n",
    "ml samtools\n",
    "ml htslib\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 8 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=4GB] \"vep -i /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique.vcf --fork 4 --cache --dir_cache /hpc/packages/minerva-centos7/vep/112/cache/ --offline --assembly GRCh38 --dir_plugins /hpc/packages/minerva-centos7/vep/112/Plugins/ --output_file /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_vep.vcf --no_stats --force_overwrite --vcf --symbol --canonical --mane_select --show_ref_allele --plugin dbNSFP,/sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP4.9a_grch38.gz,/sc/arion/projects/GENECAD/Robert/dbNSFP/dbNSFP_replacement_logic.txt,transcript_match=1,SIFT_pred,SIFT4G_pred,Polyphen2_HDIV_pred,Polyphen2_HVAR_pred,LRT_pred,MutationTaster_pred,MutationAssessor_pred,FATHMM_pred,PROVEAN_pred,MetaSVM_pred,MetaLR_pred,M-CAP_pred,PrimateAI_pred,DEOGEN2_pred,BayesDel_addAF_pred,BayesDel_noAF_pred,ClinPred_pred,LIST-S2_pred,fathmm-MKL_coding_rankscore,fathmm-MKL_coding_pred,fathmm-XF_coding_rankscore,fathmm-XF_coding_pred,Aloft_pred,VEST4_score,VEST4_rankscore,REVEL_score,REVEL_rankscore,MutPred_score,MutPred_rankscore,MVP_score,MVP_rankscore,MPC_score,MPC_rankscore,DANN_score,DANN_rankscore,CADD_raw,CADD_raw_rankscore,Eigen-raw_coding,Eigen-raw_coding_rankscore,Eigen-PC-raw_coding,Eigen-PC-raw_coding_rankscore,AlphaMissense_score,AlphaMissense_rankscore,AlphaMissense_pred,gnomAD_exomes_AF,gnomAD_exomes_POPMAX_AF,gnomAD_genomes_AF,gnomAD_genomes_POPMAX_AF\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=3GB] filter_vep -i /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_vep.vcf -o /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_vep_filter.vcf --format vcf --force_overwrite --filter \"(CANONICAL is YES) and (MANE_SELECT exists)\" --only_matched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67177968-2ee5-451f-a53e-98bf00df9899",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process VEP outputs\n",
    "vcf = pd.read_csv(f'./Finngen/Annotations/sv_unique_vep_filter.vcf', sep='\\t', skiprows=53)\n",
    "vcf = vcf.set_axis(['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO'],axis=1).drop(['QUAL','FILTER'],axis=1)\n",
    "vcf = vcf.assign(INFO=vcf['INFO'].str.split(',')).explode('INFO')\n",
    "vcf['INFO'] = vcf['INFO'].str.split('|')\n",
    "\n",
    "column_names = [\n",
    "    'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature',\n",
    "    'BIOTYPE', 'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
    "    'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'REF_ALLELE',\n",
    "    'DISTANCE', 'STRAND', 'FLAGS', 'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE_SELECT',\n",
    "    'Aloft_pred', 'AlphaMissense_pred', 'AlphaMissense_rankscore', 'AlphaMissense_score',\n",
    "    'BayesDel_addAF_pred', 'BayesDel_noAF_pred', 'CADD_raw', 'CADD_raw_rankscore',\n",
    "    'ClinPred_pred', 'DANN_rankscore', 'DANN_score', 'DEOGEN2_pred', 'Eigen-PC-raw_coding',\n",
    "    'Eigen-PC-raw_coding_rankscore', 'Eigen-raw_coding', 'Eigen-raw_coding_rankscore',\n",
    "    'FATHMM_pred', 'LIST-S2_pred', 'LRT_pred', 'M-CAP_pred', 'MPC_rankscore', 'MPC_score',\n",
    "    'MVP_rankscore', 'MVP_score', 'MetaLR_pred', 'MetaSVM_pred', 'MutPred_rankscore',\n",
    "    'MutPred_score', 'MutationAssessor_pred', 'MutationTaster_pred', 'PROVEAN_pred',\n",
    "    'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_pred', 'PrimateAI_pred', 'REVEL_rankscore',\n",
    "    'REVEL_score', 'SIFT4G_pred', 'SIFT_pred', 'VEST4_rankscore', 'VEST4_score',\n",
    "    'fathmm-MKL_coding_pred', 'fathmm-MKL_coding_rankscore', 'fathmm-XF_coding_pred',\n",
    "    'fathmm-XF_coding_rankscore', 'gnomAD_exomes_AF', 'gnomAD_exomes_POPMAX_AF',\n",
    "    'gnomAD_genomes_AF', 'gnomAD_genomes_POPMAX_AF'\n",
    "]\n",
    "\n",
    "vcf[column_names] = pd.DataFrame(vcf['INFO'].tolist(), index=vcf.index)\n",
    "vcf = vcf.drop(columns=['INFO'])\n",
    "vcf = vcf.drop(['Allele','Feature_type','Feature','BIOTYPE','EXON','INTRON','HGVSc','HGVSp',\n",
    "                'cDNA_position','CDS_position','Protein_position','Amino_acids','Codons','Existing_variation',\n",
    "                'REF_ALLELE','DISTANCE','STRAND','FLAGS','SYMBOL_SOURCE','HGNC_ID',\n",
    "                'gnomAD_genomes_AF','gnomAD_genomes_POPMAX_AF'],axis=1)\n",
    "\n",
    "vcf.loc[vcf['Consequence'].str.contains('stop_lost|start_lost|transcript_amplification|inframe_insertion|inframe_deletion|missense_variant|protein_altering_variant',na=False),'snp_type'] = 'Missense'\n",
    "vcf.loc[vcf['Consequence'].str.contains('transcript_ablation|splice_acceptor_variant|splice_donor_variant|stop_gained|frameshift_variant',na=False),'snp_type'] = 'PTV'\n",
    "vcf = vcf.loc[vcf['IMPACT'] != 'MODIFIER']\n",
    "vcf = vcf.loc[vcf['snp_type'].notna()]\n",
    "\n",
    "#####\n",
    "\n",
    "qualitative_tools = ['SIFT_pred', 'SIFT4G_pred', 'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_pred', \n",
    "                     'LRT_pred', 'MutationTaster_pred', 'FATHMM_pred', 'PROVEAN_pred', \n",
    "                     'MetaSVM_pred', 'MetaLR_pred', 'M-CAP_pred', 'PrimateAI_pred', \n",
    "                     'DEOGEN2_pred', 'BayesDel_addAF_pred', 'BayesDel_noAF_pred', \n",
    "                     'ClinPred_pred', 'LIST-S2_pred', 'fathmm-MKL_coding_pred', \n",
    "                     'fathmm-XF_coding_pred', 'MutationAssessor_pred', 'Aloft_pred']\n",
    "\n",
    "quantitative_tools = ['VEST4_rankscore', 'REVEL_rankscore', 'MutPred_rankscore', \n",
    "                      'MVP_rankscore', 'MPC_rankscore', 'DANN_rankscore', 'CADD_raw_rankscore', \n",
    "                      'Eigen-raw_coding_rankscore', 'Eigen-PC-raw_coding_rankscore']\n",
    "\n",
    "\n",
    "def calculate_score(row):\n",
    "    score = 0\n",
    "    count = 0\n",
    "    \n",
    "    # Qualitative scoring\n",
    "    for tool in qualitative_tools:\n",
    "        if tool in row and row[tool] != '':\n",
    "            if tool == 'MutationAssessor_pred' and row[tool] == 'H':\n",
    "                score += 1\n",
    "            elif tool == 'Aloft_pred' and row[tool] in ['R', 'D']:\n",
    "                score += 1\n",
    "            elif tool == 'MutationTaster_pred' and 'D' in str(row[tool]):\n",
    "                score += 1\n",
    "            elif row[tool] == 'D':\n",
    "                score += 1\n",
    "            count += 1\n",
    "    \n",
    "    # Quantitative scoring\n",
    "    for tool in quantitative_tools:\n",
    "        if tool in row and row[tool] != '':\n",
    "            try:\n",
    "                if float(row[tool]) > 0.9:\n",
    "                    score += 1\n",
    "                count += 1\n",
    "            except ValueError:\n",
    "                pass  # Ignore non-numeric values\n",
    "\n",
    "    return score, count\n",
    "\n",
    "# Apply missense scoring function\n",
    "vcf[['total_score', 'non_missing_count']] = vcf.apply(lambda row: pd.Series(calculate_score(row)), axis=1)\n",
    "\n",
    "vcf['Pred5'] = 0\n",
    "vcf.loc[vcf['SIFT_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "vcf.loc[vcf['LRT_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "vcf.loc[vcf['MutationTaster_pred'].str.contains('A|D',na=False), 'Pred5'] += 1\n",
    "vcf.loc[vcf['Polyphen2_HDIV_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "vcf.loc[vcf['Polyphen2_HVAR_pred'].str.contains('D',na=False), 'Pred5'] += 1\n",
    "\n",
    "#####\n",
    "\n",
    "vcf = vcf[['ID','CHROM','POS','REF','ALT','snp_type','Consequence','IMPACT','SYMBOL',\n",
    "           'total_score','non_missing_count','Pred5',\n",
    "           'AlphaMissense_pred','AlphaMissense_rankscore','AlphaMissense_score',\n",
    "           'REVEL_rankscore','REVEL_score',\n",
    "           'gnomAD_exomes_AF','gnomAD_exomes_POPMAX_AF']]\n",
    "vcf.to_csv(f'./Finngen/Annotations/sv_unique_vep_filter_cleaned.vcf', sep='\\t', index=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4d2766d-4c85-457e-9883-32c4a95cafbf",
   "metadata": {},
   "source": [
    "### Run LOFTEE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "855bcb4a-59b2-4009-900a-35cc210c5ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = pd.read_csv('./Finngen/Annotations/sv_unique_vep_filter_cleaned.vcf', sep='\\t')\n",
    "vcf['ID'].drop_duplicates().to_csv('./Finngen/Annotations/rare_missense_lof.txt', sep='\\t', index=False, header=None)\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9b00f420-00cd-4646-8122-f8005c3291ba",
   "metadata": {},
   "source": [
    "ml CPAN\n",
    "ml tabix\n",
    "ml vep/112\n",
    "ml perl/5.40.0\n",
    "ml samtools\n",
    "ml htslib\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 8 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=4GB] \"vep -i /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique.vcf --fork 4 --cache --dir_cache /hpc/packages/minerva-centos7/vep/112/cache/ --offline --assembly GRCh38 --dir_plugins /sc/arion/projects/GENECAD/Robert/Software/loftee-1.0.4/ --output_file /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_loftee.vcf --no_stats --force_overwrite --vcf --symbol --canonical --mane_select --show_ref_allele --plugin LoF,loftee_path:/sc/arion/projects/GENECAD/Robert/Software/loftee-1.0.4/,human_ancestor_fa:/sc/arion/projects/GENECAD/Robert/Software/human_ancestor/human_ancestor.fa.gz,conservation_file:/sc/arion/projects/GENECAD/Robert/Software/conservation_file/loftee.sql\"\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 39:59 -n 4 -o run.out -e run.error -R span[hosts=1] -R rusage[mem=3GB] filter_vep -i /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_loftee.vcf -o /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/sv_unique_loftee_filter.vcf --format vcf --force_overwrite --filter \"(CANONICAL is YES) and (MANE_SELECT exists)\" --only_matched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d0d495-192f-4a76-b7df-ec1694824f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = pd.read_csv(f'./Finngen/Annotations/sv_unique_loftee_filter.vcf', sep='\\t', skiprows=53)\n",
    "vcf = vcf.set_axis(['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO'],axis=1).drop(['QUAL','FILTER'],axis=1)\n",
    "vcf = vcf.assign(INFO=vcf['INFO'].str.split(',')).explode('INFO')\n",
    "vcf['INFO'] = vcf['INFO'].str.split('|')\n",
    "\n",
    "column_names = [\n",
    "    'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature',\n",
    "    'BIOTYPE', 'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position',\n",
    "    'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'REF_ALLELE',\n",
    "    'DISTANCE', 'STRAND', 'FLAGS', 'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE_SELECT',\n",
    "    'LoF','LoF_filter','LoF_flags','LoF_info'\n",
    "]\n",
    "\n",
    "vcf[column_names] = pd.DataFrame(vcf['INFO'].tolist(), index=vcf.index)\n",
    "vcf = vcf.drop(columns=['INFO'])\n",
    "vcf.loc[vcf['Consequence'].str.contains('stop_lost|start_lost|transcript_amplification|inframe_insertion|inframe_deletion|missense_variant|protein_altering_variant',na=False),'snp_type'] = 'Missense'\n",
    "vcf.loc[vcf['Consequence'].str.contains('transcript_ablation|splice_acceptor_variant|splice_donor_variant|stop_gained|frameshift_variant',na=False),'snp_type'] = 'PTV'\n",
    "vcf = vcf.loc[vcf['IMPACT'] != 'MODIFIER']\n",
    "vcf = vcf[['ID','Consequence','SYMBOL','LoF','LoF_filter','LoF_flags','LoF_info']]\n",
    "vcf.to_pickle('./Finngen/Annotations/vep_loftee.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14dbc4b-10c5-4457-9398-c676e06367f0",
   "metadata": {},
   "source": [
    "### Combining VEP outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a0d691bf-f8df-4565-9057-0e8433ebd6e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3075857/1911103171.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  vep = pd.read_csv('./Finngen/Annotations/sv_unique_vep_filter_cleaned.vcf', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "vep = pd.read_csv('./Finngen/Annotations/sv_unique_vep_filter_cleaned.vcf', sep='\\t')\n",
    "loftee = pd.read_pickle('./Finngen/Annotations/vep_loftee.pkl')\n",
    "vep = vep.merge(loftee, how='left')\n",
    "vep['missense_score'] = vep['total_score']/vep['non_missing_count']\n",
    "vep = vep.drop(['LoF_filter','LoF_info'],axis=1)\n",
    "\n",
    "gof = pd.read_pickle('./LoGoFunc/all_predictions.pkl')\n",
    "gof = gof.rename({'#CHROM':'CHROM'},axis=1).drop(['ID','prediction','LoGoFunc_LOF','SYMBOL'],axis=1)\n",
    "vep = vep.merge(gof, how='left')\n",
    "\n",
    "lof = pd.read_pickle('./LoGoFunc/LOF_predictions.pkl').rename({'#CHROM':'CHROM'},axis=1).drop(['ID','SYMBOL'],axis=1)\n",
    "vep = vep.merge(lof, how='left')\n",
    "\n",
    "vep['annotation'] = '5_other_missense'\n",
    "vep.loc[(vep['non_missing_count'] > 7) & (vep['missense_score'] > 0.5), 'annotation'] = '4_other_missense_0.5'\n",
    "vep.loc[vep['LoGoFunc_GOF'] > 0, 'annotation'] = '3_gof'\n",
    "vep.loc[vep['LoGoFunc_LOF'] > 0, 'annotation'] = '2_lof_lc'\n",
    "vep.loc[(vep['LoF'] == 'LC'), 'annotation'] = '2_lof_lc'\n",
    "vep.loc[(vep['LoF'] == 'HC'), 'annotation'] = '1_lof_hc'\n",
    "\n",
    "vep = vep.sort_values(['annotation','AlphaMissense_score'], ascending=[True,False]).drop_duplicates('ID')\n",
    "vep = vep[['ID','SYMBOL','CHROM','POS','REF','ALT','snp_type','Consequence','annotation',\n",
    "           'non_missing_count','missense_score']]\n",
    "\n",
    "vep.to_pickle('./Finngen/Annotations/vep_cleaned.pkl')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d11b27cf-ac28-4335-9e2d-77fa589e084b",
   "metadata": {},
   "source": [
    "### Subset common and rare variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381c2fd8-85dd-4a73-a06d-da694582501e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" \\\n",
    "'for f in /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Processed/*.sig; do awk -F\"\\t\" '\\''NR==FNR{a[$1];next}FNR==1{OFS=\"\\t\";print \"id\",\"chr\",\"pos\",\"ref\",\"alt\",\"log10p\",\"beta\",\"se\"}{id=$1\":\"$2\":\"$3\":\"$4;if(id in a)print id,$1,$2,$3,$4,$8,$9,$10}'\\'' /sc/arion/projects/GENECAD/Robert/DOE/eQTL/finngen_unique_variants.tsv \"$f\" > \"/sc/arion/projects/GENECAD/Robert/DOE/Finngen/Common/$(basename \"$f\")\"; done'\n",
    "\n",
    "bsub -q premium -P acc_GENECAD -W 23:59 -n 4 -o run.out -e run.error -R \"span[hosts=1]\" -R \"rusage[mem=1GB]\" \\\n",
    "'for f in /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Processed/*.sig; do awk -F\"\\t\" '\\''NR==FNR{a[$1];next}FNR==1{OFS=\"\\t\";print \"id\",\"chr\",\"pos\",\"ref\",\"alt\",\"log10p\",\"beta\",\"se\"}{id=$1\":\"$2\":\"$3\":\"$4;if(id in a)print id,$1,$2,$3,$4,$8,$9,$10}'\\'' /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Annotations/rare_missense_lof.txt \"$f\" > \"/sc/arion/projects/GENECAD/Robert/DOE/Finngen/Rare/$(basename \"$f\")\"; done'\n",
    "\n",
    "(echo -e \"id\\tchr\\tpos\\tref\\talt\\tlog10p\\tbeta\\tse\\tpheno\"\n",
    "for f in /sc/arion/projects/GENECAD/Robert/DOE/Finngen/Rare/finngen_R12_*.sig; do\n",
    "  pheno=$(basename \"$f\" .sig | sed 's/finngen_R12_//')\n",
    "  tail -n +2 \"$f\" | awk -v p=\"$pheno\" 'BEGIN{OFS=\"\\t\"}{print $1,$2,$3,$4,$5,$6,$7,$8,p}'\n",
    "done) > /sc/arion/projects/GENECAD/Robert/DOE/Finngen/rare_missense_lof.tsv\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
